### PR TITLE
Warnint 64to32 6186 v17

### PR DESCRIPTION
**Pull request by catenacyber**
[Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/6186

Describe changes:
- fix `-Wshorten-64-to-32` warnings for some files
- Add CI build with this warning

```
LIBHTP_BRANCH=pr/399
```
https://github.com/OISF/libhtp/pull/399

#9666 with needed rebase

There must be some stuff to review more carefully

### CODE REVIEW

File: src/app-layer-detect-proto.c [hunk 1]
```diff
@@ -686,7 +686,7 @@ static uint32_t AppLayerProtoDetectProbingParserGetMask(AppProto alproto)
         FatalError("Unknown protocol detected - %u", alproto);
     }
 
-    SCReturnUInt(1UL << (uint32_t)alproto);
+    SCReturnUInt((uint32_t)1 << (uint32_t)alproto);
```
Comment (victorjulien) [inline src/app-layer-detect-proto.c:689]: can we use `BIT_U32(alproto)`?

File: src/app-layer-htp-file.c [hunk 2]
```diff
@@ -186,13 +186,17 @@ int HTPFileOpenWithRange(HtpState *s, HtpTxUserData *txud, const uint8_t *filena
     uint8_t *keyurl;
     uint32_t keylen;
     if (tx->request_hostname != NULL) {
-        keylen = bstr_len(tx->request_hostname) + filename_len;
+        uint32_t hlen = (uint32_t)bstr_len(tx->request_hostname);
+        if (bstr_len(tx->request_hostname) > UINT16_MAX) {
```
Comment (victorjulien) [inline src/app-layer-htp-file.c:190]: should we even accept this at all? A hostname shouldn't be this long and even if it is a 64k limit seems too high?

File: src/app-layer-htp-range.c [hunk 3]
```diff
@@ -550,7 +550,13 @@ File *HttpRangeClose(const StreamingBufferConfig *sbcfg, HttpRangeContainerBlock
                     return f;
                 }
             }
-            if (FileAppendData(c->container->files, sbcfg, range->buffer, range->offset) != 0) {
+            if (range->offset > UINT32_MAX) {
```
Comment (victorjulien) [inline src/app-layer-htp-range.c:553]: should this lead to an event of some sort?

File: src/app-layer-htp-range.c [hunk 4]
```diff
@@ -571,8 +577,13 @@ File *HttpRangeClose(const StreamingBufferConfig *sbcfg, HttpRangeContainerBlock
                 }
                 // And the range ends beyond where we ended
                 // in this case of overlap, only add the extra data
-                if (FileAppendData(c->container->files, sbcfg, range->buffer + overlap,
-                            range->offset - overlap) != 0) {
+                if (range->offset - overlap > UINT32_MAX) {
```
Comment (victorjulien) [inline src/app-layer-htp-range.c:580]: event?

File: src/app-layer-htp-range.c [hunk 5]
```diff
@@ -139,7 +139,7 @@ static inline bool ContainerValueRangeTimeout(HttpRangeContainerFile *cu, const
     return false;
 }
 
-static void ContainerUrlRangeUpdate(HttpRangeContainerFile *cu, uint32_t expire)
+static void ContainerUrlRangeUpdate(HttpRangeContainerFile *cu, uint64_t expire)
```
Comment (victorjulien) [inline src/app-layer-htp-range.c:142]: if this is expire time in seconds, u32 should be enough. If it is isn't we should convert to `SCTime`

File: src/app-layer-htp-range.h [hunk 6]
```diff
@@ -40,7 +40,7 @@ typedef struct HttpRangeContainerBuffer {
     /** offset of bytes written in buffer (relative to the start of the range) */
     uint64_t offset;
     /** number of gaped bytes */
-    uint64_t gap;
+    uint32_t gap;
```
Comment (victorjulien) [inline src/app-layer-htp-range.h:43]: can we have >4GiB files, having gaps >4GiB as well?

File: src/app-layer-smtp.c [hunk 7]
```diff
@@ -519,8 +519,15 @@ int SMTPProcessDataChunk(const uint8_t *chunk, uint32_t len,
                 flags |= FILE_STORE;
             }
 
-            uint32_t depth = smtp_config.content_inspect_min_size +
-                (smtp_state->toserver_data_count - smtp_state->toserver_last_data_stamp);
+            uint32_t depth = (uint32_t)(
```
Comment (victorjulien) [inline src/app-layer-smtp.c:522]: can we do something like
```c
uint32_t depth = MIN(UINT32_t,
    smtp_config.content_inspect_min_size + (smtp_state->toserver_data_count - smtp_state->toserver_last_data_stamp));
```
Not sure if it is better.

File: src/app-layer-smtp.h [hunk 8]
```diff
@@ -124,7 +124,7 @@ typedef struct SMTPState_ {
     /** current command in progress */
     uint8_t current_command;
     /** bdat chunk len */
-    uint32_t bdat_chunk_len;
+    uint64_t bdat_chunk_len;
```
Comment (victorjulien) [inline src/app-layer-smtp.h:127]: seems like we're going to be wasting space here. Didn't immediately see a limit for BDAT chunk len. Wireshark uses u32. Haven't checked others.

File: src/defrag.c [hunk 9]
```diff
@@ -189,13 +189,13 @@ DefragContextNew(void)
 
     /* Initialize the pool of frags. */
     intmax_t frag_pool_size;
-    if (!ConfGetInt("defrag.max-frags", &frag_pool_size) || frag_pool_size == 0) {
+    if (!ConfGetInt("defrag.max-frags", &frag_pool_size) || frag_pool_size == 0 ||
```
Comment (victorjulien) [inline src/defrag.c:192]: can we use a size parsing func with valid range specified?

File: src/defrag.h [hunk 10]
```diff
@@ -37,7 +37,7 @@ typedef struct DefragContext_ {
     Pool *frag_pool; /**< Pool of fragments. */
     SCMutex frag_pool_lock;
 
-    time_t timeout; /**< Default timeout. */
+    int timeout; /**< Default timeout. */
```
Comment (victorjulien) [inline src/defrag.h:40]: should be unsigned as negative as no meaning

**Review by victorjulien — CHANGES_REQUESTED** [review]
Partial review of the first 27 files before running out of time.

File: src/defrag.h (hunk 10 above)
Comment (catenacyber) [reply src/defrag.h:40]: ok

File: src/defrag.c (hunk 9 above)
Comment (catenacyber) [reply src/defrag.c:192]: `ConfGetInt` has no variant with range...
Using `StringParseU64RangeCheck` would need more lines of code : first get the string from the config, then parse it and check its range...

This looks simpler, does it not ?

File: src/app-layer-smtp.h (hunk 8 above)
Comment (catenacyber) [reply src/app-layer-smtp.h:127]: Fixing by using `StringParseUint32` instead of `strtoul`

File: src/app-layer-htp-range.h (hunk 6 above)
Comment (catenacyber) [reply src/app-layer-htp-range.h:43]: If a TCP stream has multiple gaps, whose addition is over 4GiB...

What do you think ? We should keep it u32 and saturate it ?

File: src/app-layer-htp-range.c (hunk 5 above)
Comment (catenacyber) [reply src/app-layer-htp-range.c:142]: @jlucovsky what do you think about this ?
This comes from `SCTIME_SECS` returning a u64

File: src/app-layer-detect-proto.c (hunk 1 above)
Comment (catenacyber) [reply src/app-layer-detect-proto.c:689]: Nope, gives a warning
```
  CC       app-layer-detect-proto.o
app-layer-detect-proto.c:689:18: warning: implicit conversion loses integer precision: 'unsigned long' to 'uint32_t' (aka 'unsigned int') [-Wshorten-64-to-32]
  689 |     SCReturnUInt(BIT_U32(alproto));
```

File: src/app-layer-htp-file.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-htp-file.c:190]: What would be the good limit ?

File: src/app-layer-detect-proto.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-detect-proto.c:689]: the macro would need the same fix

File: src/app-layer-htp-file.c (hunk 2 above)
Comment (victorjulien) [reply src/app-layer-htp-file.c:190]: isn't there a 256 limit on hostnames? Don't recall with certainty

File: src/app-layer-htp-range.c (hunk 3 above)
Comment (victorjulien) [reply src/app-layer-htp-range.c:553]: not yet, but there should be a tx close?

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .uptime | 939 | 992 | 105.64% |
| .decoder.event.ipv4.frag_overlap | 805 | 511 | 63.48% | 

Pipeline 16585

**Comment by victorjulien** [comment]
> WARNING:
> 
> field	baseline	test	%
> SURI_TLPR1_stats_chk		
> .uptime	939	992	105.64%
> .decoder.event.ipv4.frag_overlap	805	511	63.48%
> Pipeline 16585

I wouldn't expect side effects from a PR like this. If there are proper fixes or improvements, please put those in a separate PR.

File: src/app-layer-htp-file.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-htp-file.c:190]: https://curl.se/docs/CVE-2023-38545.html :-p

File: src/app-layer-detect-proto.c (hunk 1 above)
Comment (catenacyber) [reply src/app-layer-detect-proto.c:689]: indeed BIT_U16 macro seems to have a fix already

**Comment by catenacyber** [comment]
> 
> I wouldn't expect side effects from a PR like this. If there are proper fixes or improvements, please put those in a separate PR.

Neither did I.

Let's see if QA confirms deviation for defrag in next PR... https://github.com/OISF/suricata/pull/9840

**Pull request closed without merging**